### ISSUE TITLE
Function to get all neighbors in World?

### ISSUE THREAD

**Issue by emilydolson** [devosoft/Empirical#237]
Currently, Worlds have a member variable (`fun_get_neighbor`) that holds a function that returns a random neighbor. There are also a lot of situations where it would be useful to have all of the neighbors of cell, rather than just one. 

Should we also add a `fun_get_neighbors` member variable? Alternatively, we could replace `fun_get_neighbor` with `fun_get_neighbors` and anyone who just wants a single random neighbor can choose randomly from what that function returns (I assume the reason we don't currently do this is because having to return all of the neighbors could be inefficient?). Thoughts?

**Comment by mmore500** [comment]
Could we provide a variant where we return a single neighbor specified via argument (e.g., direction) ? We would need a direction specifying scheme that would be consistent as we increase dimensionality (e.g. 2d vs 3D) which should be possible but would require some thought. Perhaps then the caller could iterate through the directions instead of packing up a container with all the neighbors?

**Comment by mmore500** [comment]
It would be useful on its own to be able to grab a neighbor by direction.

**Comment by emilydolson** [comment]
I really like that idea, but I have two concerns:
- In some cases, calculating neighbors in a direction might end up being substantially more expensive than calculating neighbors (I'm imagining trying to divide a sphere into sectors and find the members of each sector or something)
- I'm worried that a universally meaningful direction scheme isn't possible. Although there are ways to handle scaling up dimensions, I suspect things will get pretty ugly if people are using some sort of weird network topology as their population structure.

**Comment by mmore500** [comment]
Ah! I had assumed a grid-type layout. Perhaps in a non-grid layout the "direction" wouldn't specify anything that is meaningful but would could stand in for unique identity of the neighbor (e.g., `0` through `n-1` where `n` is number of neighbors). Under the hood, we would just store a vector of neighbors and for grid-based situations the construction process would fill the neighbor vector in a direction-meaningful manner while in a more customized situation the user would just fill the neighbor vector as desired (possibly arbitrarily). Would iterating through neighbors in an arbitrary order up through num neighbors meet your use case requirement?

**Comment by mmore500** [comment]
Are you assuming that neighbors would be defined in terms of each individual in the population or you would have neighbor set A and neighbor set B and an individual would either be in A or B?

**Comment by emilydolson** [comment]
Iterating through neighbors in an arbitrary order works for me! The thing I'm doing right now is a Cellular Automaton variant, so I need to look at all of the neighbors to follow the update rule. I know I've thought about this before in other contexts too, but I think a random order would be fine for all of them (since they all just required an unordered set of neighbors).

I can imagine situations where either defining neighbors relative to an individual or by groups would be the best option (which is why I like the flexibility of just having the user write a function).

**Comment by anyaevostinar** [comment]
On a more basic level, it would be great if there were presets for the common neighbor functions liked mixed and grid, or is this a thing already?

**Comment by emilydolson** [comment]
There are presets for mixed and grid (possibly a few others?) but they're all in the current format, which returns a single neighbor rather than a vector of all of them. Which is useful in a lot of contexts (placing offspring, etc.) but sometimes you really need them all.

**Comment by mmore500** [comment]
we could potentially use MatchBin for this... where the bin contains "tags" (position specifiers) of every position in the grid... the distance metric is however you want to define distance on your spatial layout (e.g., euclidean for grid, equal distance for mixed)... and then you can use the ranked selector to get the best n neighbors... there's built-in caching so the computations would only need to take place once

**Comment by emilydolson** [comment]
This is partially solved by #402, which adds `GetValidNeighborOrgIDs`. This is an inefficient solution for situations where all valid neighbor ids could be calculated directly (rather than looping through the entire population, which is what happens in `GetValidNeighborOrgIDs`), but it does work. Not sure whether it's worth also having a user-defined function to do this or not.

**Comment by anyaevostinar** [comment]
Agreed, probably good to improve the efficiency of GetValidNeighborOrgIDs in situations where it can be calculated directly
